List all commits for integration

cojoj opened the pull request:
Working now on **List all commits for integration** models.

czechboy0 commented:
Looking good so far!

cojoj commented:
I wonder what `status` is in:
```json
"XCSCommitCommitChangeFilePaths": [{
					"status": 4,
					"filePath": "XcodeServerSDK.podspec"
				}],
```

czechboy0 commented:
Changed, added or deleted? Maybe?

czechboy0 commented:
And moved?

cojoj commented:
Yeah, you're right...

So far I got this:

0. Dunno
1. **Added**
2. Dunno
3. Dunno
4. **Changed/Modified**

cojoj commented:
Do you think that this is documented somewhere? MongoDB doesn't hold types but from reverse engineering you can see that this will be be based on bit modifications.

czechboy0 commented:
:+1:

czechboy0 commented:
And btw I checked everywhere and I can't find where all the file change statuses are defined.

cojoj commented:
I guess it must be hardcoded somewhere so we won't get access to this... 😢

czechboy0 commented:
I don't think it's intentional, tbh. Anyway, put it in an `enum` with `Int` raw value.

cojoj commented:
So far, I got this
```swift
public enum FileStatus: Int {
    case Added = 1
    case Deleted = 2
    case Modified = 4
    case Moved = 8192
}
``` 

I guess it isn't possible to call something like:
```swift
FileStatus(rawValue: 8)
```

So we should make it `Optional`?

cojoj commented:
Came with this. What do you think @czechboy0?
```swift
public enum FileStatus: Int {
    case Added = 1
    case Deleted = 2
    case Modified = 4
    case Moved = 8192
    case Other
    
    public init?(rawValue: Int) {
        switch rawValue {
        case 1:
            self = .Added
        case 2:
            self = .Deleted
        case 4:
            self = .Modified
        case 8192:
            self = .Moved
        default:
            self = .Other
        }
    }
}
```

czechboy0 commented:
Don't enums already have failable initializers generated by default?

czechboy0 commented:
So in the Commit you'd do `let fileStatus = FileStatus(rawValue: parsed) ?? .Other`

cojoj commented:
Yes, they have but I was hoping I can somehow assign other value to `.Other` but I guess it's impossible...

![](http://media1.giphy.com/media/xTiTntReleqBnhBNwQ/giphy.gif)

czechboy0 commented:
Yeah I'm afraid `enum`s don't have a `default` value, which is basically what you're looking for. Radar?

cojoj commented:
It's a bit controversial... Should `enum`s  have default values? But on the other hand, they already store `rawValues` so where's the problem to assign `.Other` value?

czechboy0 commented:
Well, they have a failable initializer. It'd be nice to have the `enum`, instead of potentially returning `nil`, when the `rawValue` doesn't match a case, return one of the predefined cases (the `default` case). Something like Hash default values in Ruby, which you can tell to return a specific value when you're trying to pull a value for an unknown key.

This would be nice because you could have the enum have a non-failable initailizer, because it'd *always* return one of the cases.

cojoj commented:
Exactly but that's probably why failable initializers are default ones in `enum`s...

cojoj commented:
@czechboy0 we have an answer straight from  about our little `enum` thing:

```swift
public enum FileStatus: RawRepresentable {
  case Added, Deleted, Modified, Moved, Other(Int)

  init(rawValue: Int) {
    switch rawValue {
    case 1: self = Added
    case 2: self = Deleted
    case 4: self = Modified
    case 8192: self = Moved
    default: self = Other(Int)
  }

  var rawValue: Int {
    switch self {
    case Added: return 1
    case Deleted: return 2
    case Modified: return 4
    case Moved: return 8192
    case Other(let x): return x
    }
  }
}
```

czechboy0 commented:
Hmm interesting, but is all this code worth it? I still don't hate `let fileStatus = FileStatus(rawValue: parsed) ?? .Other` as much I'd hate to have this much code to achieve the same thing :(

cojoj commented:
The thing I keep in mind when thinking about this is **we only know 4 values for `status`**. I want to keep somehow value of `status`es which aren't known yet... ← This is my personal opinion.

On the other hand - this will be an official API call so there _should_ be a documentation for this soon (they've promised beta 3 or 4 but still not here...). If we get ths documentation we can easily add missing cases and live happily after all 😉

I hate making decisions... So, it's up to you 😜

czechboy0 commented:
Yeah I get your point. But IMHO, if we do see `.Other` state somewhere, we can always debug and take a look at the actual value. Since we don't need the value anywhere else (at the moment), I would only see this as a small help during debugging. Which, as you said, hopefully should not be required if Apple gives us official docs on this. 

I'd go with the simpler solution for now.

cojoj commented:
Fair enough! 👍

buildasaur commented:
Result of Integration **1**
---
*Duration*: 23 seconds
*Result*: **-1error, failing state: build-failed**

buildasaur commented:
Result of Integration **1**
---
*Duration*: 28 seconds
*Result*: **4errors, failing state: build-errors**

czechboy0 commented:
Can you please merge `czechboy0/swift-2` into this branch? I made changes for Xcode7b4 compatibility.

cojoj commented:
Done with merging...

![](http://media.giphy.com/media/Mp4hQy51LjY6A/giphy.gif)

cojoj commented:
`results` contain array... In our case it'll be of type `[IntegrationCommits]`. Do you think thath it's possible there'll be more than one element in this `results` array? If yes then I think name of this class should be changed... If no, we should return only one object of type `IntegrationCommits`.

cojoj commented:
This is my implementation of `init` - just code, without any proof... We'll have to record a cassete and write `request` method for this. I still think that the name `commits` in this JSON is confusing, so I just `flatMap`ped this shit... 💩

buildasaur commented:
Result of Integration **1**
---
*Duration*: 21 seconds
*Result*: **-1 error, failing state: build-failed**

buildasaur commented:
Result of Integration **2**
---
*Duration*: 51 seconds
*Result*: **Perfect build!** All 36 tests passed. :+1:
*Test Coverage*: 48%.

buildasaur commented:
Result of Integration **2**
---
*Duration*: 1 minute and 4 seconds
*Result*: **Perfect build!** All 36 tests passed. :+1:
*Test Coverage*: 48%.

cojoj commented:
Do you really think it'd be logical to have a `Dictionary` instead of `Array` as `commits` property? Imagine yourself using this... 
> "I just want to see how many commits have this integration"
> "WTF, why it's a Dictionary of freaking `Strings` instead of `[Commit]`!?"

That's why I'm using here `flatMap`, so when I access `commits` object in JSON and get `Dictionary` of  type `[String: [Commit]]` from it I'm taking `allValues`, so the array of arrays and `flatten` it so we'll have one dimensional array. Why I'm `flatMap`ping? As we've discussed - we may have more `RepositoryBlueprintID`s (which we don't know how to prove, yet).

czechboy0 commented:
I think a repository with git submodules will have multiple repository ids. 

But yes, I think we should map the returned objects as closely as possible. The user can always do the flat map themselves, but right now we're taking away information about which repo generated those commits. 

The purpose of this project is to map the JSON api into Swift objects. Plus we have a couple of utility functions. But we shouldn't be dictating what information the user doesn't need to see, if we have it. That's up to them to decide. So yes, I think the type should be preserved from the JSON, `[String: [Commit]]`.

cojoj commented:
From my POV, Apple screwed naming... Why `commits`? 😑

czechboy0 commented:
Yeah maybe. `commits` as in this is where you get commits. But they are keyed by repo. What else would you name it?

cojoj commented:
Now we can't do anything... But when they post docs for this endpoints and we'll be assured that those are submodules, we can name it `submodules`...

cojoj commented:
Got any pattern on how to work with conversion from `NSDictionary → Dictionary`?

Also, please review in free time this [pull request](https://github.com/buildasaurs/BuildaUtils/pull/1) as it's some kind of _blocker_ for me 😓

czechboy0 commented:
1. I merged you PR in BuildaUtils, in the Podfile change the BuildaUtils version to `0.0.6`.
2. Just do an optional cast from `NSDictionary` to your type of `Dictionary`. It will check the objects inside to make sure it contains the right objects. Is that what you mean?

buildasaur commented:
Result of Integration **3**
---
*Duration*: 44 seconds
*Result*: **Perfect build!** All 37 tests passed. :+1:
*Test Coverage*: 49%.

cojoj commented:
> Lately I changed my opinion and now I think we should really minimize the places where we force unwrap optionals.

![](http://www.awyisser.com/assets/images/thumbnail.png)

Now, change your opinion and let's drop all those `NSDictionaries` 😍

---
I will try it later... Now **it's pizza Friday**!!! 🍕

czechboy0 commented:
How do you mean? Where do you think we shouldn't be using `NSDictionary`s?

cojoj commented:
Call me hipster or whatever but I really hate mixing Swift with Objective-C... For me, this project is totally platform independant (we don't use any of UI stuff specific to OS) that's why I see a possibility to _do it Swifty_.

I know that parsing JSON without `NSDictionary` will be a PITA but. hey, it's Open Source, we have tame, we have possibilities and we're willing to learn new stuff, right?

Because of these _easy_ casting to `NSDictionary` or `NSArray` we're fighting with Swift's **static typing** and (I think so) losing some security.

---
It's still your project and I'm only, a contributor so... **I'll do it by following your leads** but I only wanted to share my honest opinion 😉

czechboy0 commented:
In principle I agree, but JSON is a special case IMHO. Even if we have a native `JSON` object in Swift, it will still be dynamic when it comes to its contents, because it can contain any sorts of stuff. So we won't get any added security because the types are not known at compile time.

I agree it would be nice to get rid of all `NS` classes, but in practice I'm not sure I understand what you're proposing :)

cojoj commented:
You probably won't exactly understand what I mean as this is only _me talking_, so don't bother 😉

I hate that we have this Objective-C **technical debt** in every code we write right now... 😭

czechboy0 commented:
Yeah I feel you, but it could be *much* worse, trust me. Plus, most of Apple's energy and time seems to be poured into Swift now. There is hope :wink:

czechboy0 commented on XcodeServerSDK/API Routes/XcodeServer+Integration.swift:
```diff
@@ -161,6 +161,38 @@ extension XcodeServer {
+                return
+            }
+            
+            let integrationCommits = IntegrationCommits(json: integrationCommitsBody[0] as! NSDictionary)
```
Why are we taking only the first `integrationCommitsBody` item?

cojoj replied:
Great question... I was asking myself the same one yesterday while writing this.

It's an `Array` but we don't know of this array will return more than one element - we know similiar things happened earlier.
**If this can return more elements than we should change the name!**

(I need a shrug emot shortcut... Just like @esttorhe)

czechboy0 replied:
Ah I see. Okay then, don't worry about it.

czechboy0 commented on XcodeServerSDKTests/IntegrationTests.swift:
```diff
@@ -23,4 +23,19 @@ class IntegrationTests: XCTestCase {
+        let server = self.getRecordingXcodeServer("get_integration_commits")
+        server.getIntegrationCommits("56ad016e2e3993ca0b8ed276050150e8") { (commitIntegrations, error) in
+            XCTAssertNil(error, "Error should be nil")
+            
```
If you could add a couple of assertions about the passed-in `commitIntegrations`'s values, so that we verify the parsing went well, that'd be great. (If you were already planning to do this, ignore me)

cojoj replied:
I'm ignoring you... 😜

czechboy0 replied:
:+1:

czechboy0 commented on XcodeServerSDK/Server Entities/IntegrationCommits.swift:
```diff
@@ -0,0 +1,42 @@
+            let blueprintID = key as! String
+            let commitsArray = (value as! NSArray) as Array
+            
+            resultsDictionary[blueprintID] = commitsArray.map { Commit(json: $0 as! NSDictionary) }
```
Yeah let's be a bit gentler with force unwrapping - I mentioned this earlier :)

cojoj replied:
I'll try but I can't promise anything! 😂

czechboy0 replied:
https://github.com/cojoj/XcodeServerSDK/commit/2cde22d668b7f81f127a0c535d4645e881647bdb#commitcomment-12340549

czechboy0 commented:
Just a couple of comments I added - awesome stuff as usual, @cojoj! Thanks for all your work here :+1:

czechboy0 commented:
Let me know when this is ready, @cojoj! :wink:

cojoj commented:
Probably tomorrow as I don't know if I'll have time today to do some coding... 😕

buildasaur commented:
Result of Integration **4**
---
*Duration*: 53 seconds
*Result*: **Perfect build!** All 56 tests passed. :+1:
*Test Coverage*: 54%.

buildasaur commented:
Result of Integration **5**
---
*Duration*: 48 seconds
*Result*: **Perfect build!** All 56 tests passed. :+1:
*Test Coverage*: 54%.

cojoj commented:
Casting fixed (looks neat). I'll now spend a while with writing some test cases and after that I think we'll be ready with this.

buildasaur commented:
Result of Integration **6**
---
*Duration*: 57 seconds
*Result*: **Perfect build!** All 56 tests passed. :+1:
*Test Coverage*: 54%.

cojoj commented:
### It works!

![](http://media4.giphy.com/media/pRxj9REqNYPUQ/giphy.gif)

Ready to review 👌

czechboy0 commented:
looking good!

cojoj commented:
I've commented some code as I'm not fully happy with it... Please take a look in some spare time. Maybe together we can work out something better 👌

buildasaur commented:
Result of Integration **7**
---
*Duration*: 53 seconds
*Result*: **Perfect build!** All 56 tests passed. :+1:
*Test Coverage*: 54%.

czechboy0 commented:
You can use the new BuildaUtils spec 0.0.7

cojoj commented:
Looks like everything's fine now 😉 Take a look!

buildasaur commented:
Result of Integration **8**
---
*Duration*: 54 seconds
*Result*: **Perfect build!** All 56 tests passed. :+1:
*Test Coverage*: 54%.

buildasaur commented:
Result of Integration **9**
---
*Duration*: 50 seconds
*Result*: **Perfect build!** All 56 tests passed. :+1:
*Test Coverage*: 54%.

czechboy0 commented:
Awesome! Thanks @cojoj, great stuff.

czechboy0 commented:
Our code coverage is sneaking up, great job everyone! :+1:
